### DISCUSSION TITLE
Ability to choose Database for History Log

### DISCUSSION THREAD

**Discussion by theAkito** [nushell/nushell#7239] [Ideas]
I was researching in this repository, whether anyone requested a different method of saving CLI history, beyond `sqlite3`.
To my surprise, I did not find any such request.

It'd be great, if there were a Plugin(?) API, which allows contributing support for different database platforms.

If I'm not mistaken `plaintext` & `sqlite3` are the two hard-coded options, that may be chosen now, without the ability to have another rather arbitrary built-in option available.

**Comment by fdncred** [comment]
You are correct, plaintext and sqlite are the two modes that our line editor, reedline, works in right now. There is not an option for mysql/postgres/redis/etc for history storage. I'm sure reedline could be extended to support that but it would take a dedicated contributor to add that functionality.

**Reply by theAkito** [reply]
Is it technically possible, as of right now, to add a generic interface for adding an arbitrary database platform?

**Reply by fdncred** [reply]
(nearly)Anything is technically possible via changing the rust code in reedline and nushell. It just takes someone with a good understanding of that area to commit to implementing such an interface.

**Reply by theAkito** [reply]
Of course, but it's usually a big difference, whether one is implementing a huge new thing, that must be designed properly, etc. as opposed to hooking onto something already existing and just extending what already is there, with the rules & restrictions, which are already there. This also already makes sure, there is no debate about the feature in the first place.
With brand new features, it's not even clear, if they are welcome, in the first place.

To clarify my earlier question:
How resource consuming would it be to add such an interface? For example, if the current two options are all about being hard-coded, without any generic interfaces, then one would need to implement the whole thing from scratch.

**Reply by fdncred** [reply]
New features like this are welcome, assuming they have been discussed in Discord in #design-discussion, and approved first.

My guess is that this would be a ~7 implementation, on a range of 1 = trivial and 10 = massive. It would definitely take more tooling and something that communicates with multiple database, perhaps something like sqlx. I believe sqlx is async which may push this to an 8 or so because I'm not sure of how async reedline is. Of course, we wouldn't want something hard coded. We'd probably want trait(s) that could be extended to provide generalized functionality across different db destinations and probably want it behind a feature. Those are some of the type of things that we'd need to discuss further in discord.

**Comment by c4lliope** [comment]
@theAkito - https://atuin.sh is a reliable choice across shells - their nushell bindings are seamless, and you can expand to combine the sessions from many machines.
